consensus/src/processes/pruning.rs

```diff
@@ -169,6 +180,86 @@ impl<S: GhostdagStoreReader, T: ReachabilityStoreReader, U: HeaderStoreReader, V
         let pp_bs = self.headers_store.get_blue_score(pruning_point).unwrap();
         pov_blue_score >= pp_bs + self.pruning_depth
     }
+
+    pub fn is_valid_pruning_point(&self, pp_candidate: Hash) -> bool {
+        if pp_candidate == self.genesis_hash {
+            return true;
+        }
+
+        let hst = self.header_selected_tip_store.read().get().unwrap().hash;
+
+        if !self.reachability_service.is_chain_ancestor_of(pp_candidate, hst) {
+            return false;
+        }
+
+        let hst_bs = self.ghostdag_store.get_blue_score(hst).unwrap();
+        self.is_pruning_point_in_pruning_depth(hst_bs, pp_candidate)
+    }
+
+    pub fn are_pruning_points_in_valid_chain(&self, pruning_info: PruningPointInfo) -> bool {
+        // We want to validate that the past pruning points form a chain to genesis. Since
+        // each pruning point's header doesn't point to the previous pruning point, but to
+        // the pruning point from its POV, we can't just traverse from one pruning point to
+        // the next one by merely relying on the current pruning point header, but instead
+        // we rely on the fact that each pruning point is pointed by another known block or
+        // pruning point.
+        // So in the first stage we go over the selected chain and add to the queue of expected
+        // pruning points all the pruning points from the POV of some chain block. In the second
+        // stage we go over the past pruning points from recent to older, check that it's the head
+        // of the queue (by popping the queue), and add its header pruning point to the queue since
+        // we expect to see it later on the list.
+        // The first stage is important because the most recent pruning point is pointing to a few
+        // pruning points before, so the first few pruning points on the list won't be pointed by
+        // any other pruning point in the list, so we are compelled to check if it's refereced by
+        // the selected chain.
+        let mut expected_pps_queue = VecDeque::new();
+        let hst = self.header_selected_tip_store.read().get().unwrap().hash;
+        for current in self.reachability_service.backward_chain_iterator(hst, pruning_info.pruning_point, false) {
+            let current_header = self.headers_store.get_header(current).unwrap();
+            if expected_pps_queue.is_empty() || *expected_pps_queue.back().unwrap() != current_header.pruning_point {
```

Comment:
`if expected_pps_queue.back().is_none_or(|&&h| h != current_header.pruning_point)` is a bit more readable